### ISSUE TITLE
[css-backgrounds-4] Split CSS Backgrounds into separate specs?

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#7664]
The CSS Backgrounds and Borders Module Level 3 is already pretty heavyweight. Level 4 adds even many more features which evolve at different speeds.

Therefore, I suggest to split the specification up into two separate ones. The first one - CSS Backgrounds Module Level 4 - focuses on only background-related features. The second one - CSS Borders Module Level 4  - covers everything related to border styling including box shadows.

While backgrounds and borders are related and interact to some extent, they still cover individual concepts.
So the two specs could focus on one main feature.

The hope behind this idea is also to let them advance more quickly.

Sebastian

**Comment by SebastianZ** [comment]
@fantasai @LeaVerou @bert-github What do you think?

**Comment by LeaVerou** [comment]
From a quick look, it appears that Borders would basically contain everything, except the Backgrounds section. Not sure that would help too much?

**Comment by SebastianZ** [comment]
The features list for Borders is currently longer than the one for Backgrounds, though Backgrounds would still contain a bunch of additions. Here's a list to make it more obvious:

Backgrounds:

* Physical and logical `background-position-*` longhands
* Optional offset(s) in `background-position-*` properties
* Logical keywords for `background-position-*` properties
* `background-clip`
* [`extend` keyword for `background-repeat`](https://lists.w3.org/Archives/Public/www-style/2010Dec/0327.html)

Borders:

* `corner-shape`
* Multiple colors per border (via [`stripes()` function](https://github.com/w3c/csswg-drafts/pull/7029))
* [Physical and logical `border-*-radius` shorthands](https://github.com/w3c/csswg-drafts/pull/7705)
* Logical `border-*` properties (maybe; currently they're [defined in CSS Logical Properties 1](https://drafts.csswg.org/css-logical-1/#border-properties))
* [`box-shadow-*` longhands](https://github.com/w3c/csswg-drafts/pull/6083)
* Split-up values for horizontal and vertical `box-shadow` spread radius

Sebastian

**Comment by SebastianZ** [comment]
I think this might be a quick decision. So let's discuss this on a call.

Sebastian

**Comment by LeaVerou** [comment]
I’m generally supportive of more modularization and this sounds like a good one (little interdependency between features).
Yes, this means I disagree with myself from Sep 9 😛

**Comment by fantasai** [comment]
Another option would be to split out into https://www.w3.org/TR/css-box-4/
In that case I might suggest keeping border-image and box-shadow with the rest...

**Comment by astearns** [comment]
Is there a current advancement opportunity for Backgrounds or Borders that this split would enable? I’m happy to take on the work if it’s immediately useful, but am less interested if it’s just a theoretical way to move through the spec stages faster

@SebastianZ are you volunteering to do the work?

**Comment by SebastianZ** [comment]
> @SebastianZ are you volunteering to do the work?

Yes, if we decide to split them, it's my plan to make that change.

> Is there a current advancement opportunity for Backgrounds or Borders that this split would enable?

It is one big goal for me this year to push the features related to backgrounds and borders. So we can finally get rid of the big "Not Ready For Implementation" banner and at least publish them as a WD.

There is currently some momentum related to some features of the spec. and I believe that a split could help making progress on them, make it easier to maintain them. And two smaller specs are also easier to read than a monolytic one, especially with that many features planned.
At least from the discussions lately, there's definitely some implementer interest and I hope this grows stronger once there's no repelling banner anymore. And that banner can be dropped individually when there are two specs.

Sebastian

**Comment by SebastianZ** [comment]
@fantasai It makes sense to split the things related to the box model (i.e. `border-*-width`) out into CSS Box Model 4. (And I saw, you already mentioned that in the abstract of that spec. 😃)

So, yeah, that could be considered when splitting the specs.

Sebastian

**Comment by LeaVerou** [comment]
If Backgrounds & Borders becomes just "Backgrounds" though, `box-shadow` and friends doesn't fit in it any more, but not sure it fits in css-box either… Maybe we need a box-decoration spec…

**Comment by SebastianZ** [comment]
> If Backgrounds & Borders becomes just "Backgrounds" though, `box-shadow` and friends doesn't fit in it any more, but not sure it fits in css-box either… Maybe we need a box-decoration spec…

It seems my comment was misleading. Backgrounds & Borders would still be split into a "Backgrounds" and a "Borders" spec. The only thing that'd be moved to CSS Box 4 would be the `border-*-width` properties.

Sebastian

**Comment by LeaVerou** [comment]
So there would be a Borders spec that doesn't include the `border-*` properties?! That seems *very* weird.

**Comment by SebastianZ** [comment]
> So there would be a Borders spec that doesn't include the `border-*` properties?!

No, it would include all `border-*` properties like `border-*-style`, `border-*-color`, `border-*-radius`, etc. Only the `border-*-width` properties might be moved to CSS Box 4, because those _also_ relate to the box model and fit to `padding-*` and `margin-*`.

Though it was @fantasai's proposal to split out parts to CSS Box 4, not mine. So it might be that I misunderstood her idea.

I'd say, let's keep this issue focused on splitting Backgrounds and Borders into two separate specs as initially intended and then have a separate discussion on whether to move parts of Borders to CSS Box 4.

Sebastian

**Comment by Loirooriol** [comment]
IMO `box-shadow` fits better into Backgrounds than Borders.

**Comment by SebastianZ** [comment]
For what it's worth, I believe it doesn't quite fit into both of them. Though details like that can still be discussed when we agreed to do the split.

Sebastian

**Comment by LeaVerou** [comment]
> IMO `box-shadow` fits better into Backgrounds than Borders.

If it gets renamed to something else, like "Backgrounds & Box decorations" or something. Otherwise it doesn't really fit into any of them. As it currently stands, Backgrounds & Borders implicitly spans more widely than both backgrounds or borders (sort of implies "and everything in between"), if split the resulting scopes sound narrower.

**Comment by SebastianZ** [comment]
Removing the resolution proposal label for now, as there's obviously still need for discussion.

@LeaVerou @Loirooriol So, what should we do with `box-shadow` (and it's [longhands](https://github.com/w3c/csswg-drafts/pull/6083))?

Possible solutions I see are:

1. Include them in the Borders spec.
2. Include them in the Backgrounds spec.
3. Move them into their own Box Decorations spec.

Thinking about it, I now slightly tend to option 3. As mentioned before, box shadows actually neither fit to borders nor to backgrounds. And we might add other box decorations beyond shadows in the future.

Sebastian

**Comment by SebastianZ** [comment]
Having a separate spec. for box decorations also goes well will the [spec. we have for text decorations](https://drafts.csswg.org/css-text-decor/).

So my suggestion is now to split the current spec. into three:

- CSS Backgrounds 4
- CSS Borders 4
- CSS Box Decorations 4

Sebastian

**Comment by SebastianZ** [comment]
As I can see some support from different people, I put this back on the agenda.

The proposal now is to split CSS Backgrounds 4 into three specs as noted in [my previous comment](https://github.com/w3c/csswg-drafts/issues/7664#issuecomment-1464704443). How exactly they are split is [outlined in a previous comment](https://github.com/w3c/csswg-drafts/issues/7664#issuecomment-1241872613) with the addition that everything `box-shadow` related will be moved to CSS Box Decorations 4. Unless expressed otherwise, the current editors will be added to the new specs as well.

Sebastian

**Comment by SebastianZ** [comment]
@fantasai You mentioned `box-decoration-break` several times in the call and we also resolved to move it to CSS Box. Though that property is currently [part of CSS Break](https://drafts.csswg.org/css-break/#break-decoration), so unrelated to the split of CSS Backgrounds, actually.

I assume you got confused by the suggested Box Decorations spec., which we resolved to _not_ add for now. Right? Does this need another resolution or can we just inore that part and stick with the rest of the resolution.

Sebastian
